lib/plugin-api/hydrogen-kernel.js

```diff
@@ -1,15 +1,65 @@
 /* @flow */
 
 import type Kernel from "./../kernel";
-import type ZMQKernel from "./../zmq-kernel";
+
+// We define all our own types here to isolate the plugin API from internal
+// implementation details of hydrogen.
+type HydrogenResultsCallback = (
+  message: any,
+  channel: "shell" | "iopub" | "stdin"
+) => void;
+
+interface HydrogenKernelMiddlewareStack {
+  interrupt: () => void;
+  shutdown: () => void;
+  restart: (onRestarted: ?Function) => void;
+  execute: (
+    code: string,
+    callWatches: boolean,
+    onResults: HydrogenResultsCallback
+  ) => void;
+  complete: (code: string, onResults: HydrogenResultsCallback) => void;
+  inspect: (
+    code: string,
+    cursorPos: number,
+    onResults: HydrogenResultsCallback
+  ) => void;
+}
+
+interface HydrogenKernelMiddleware {
```

Comment:
Why do we need `HydrogenKernelMiddleware` and `HydrogenKernelMiddlewareStack`?

Reply:
Say a plugin defines `execute(next, ...)` which calls `next.execute`. Note that the first argument of `next.execute` is *not* `next`, so it need to have a different type.